lib/command.js

```diff
@@ -229,17 +230,35 @@ module.exports = function command (yargs, usage, validation, globalMiddleware) {
 
     if (commandHandler.handler && !yargs._hasOutput()) {
       yargs._setHasOutput()
-      if (commandHandler.middlewares.length > 0) {
-        const middlewareArgs = commandHandler.middlewares.reduce(function (initialObj, middleware) {
-          return Object.assign(initialObj, middleware(innerArgv))
-        }, {})
-        Object.assign(innerArgv, middlewareArgs)
-      }
-      const handlerResult = commandHandler.handler(innerArgv)
-      if (handlerResult && typeof handlerResult.then === 'function') {
-        handlerResult.then(
-          null,
-          (error) => yargs.getUsageInstance().fail(null, error)
+
+      const middlewares = commandHandler.middlewares || []
+
+      innerArgv = middlewares
+        .reduce((accumulation, middleware) => {
+          if (isPromise(accumulation)) {
```

Comment:
Ideally we would follow the same naming conventions as Node and expose two methods `middleware` and `middlewareSync`.  That way we save on installing `isPromise` and running that check `middlewares.length` times.

Would have been great to have realized this when we added this to yargs api.